Selectable table

keller-mark opened the pull request:
Towards #491 

This pull request adds a relatively general `<SelectableTable />` component.
It also adds a wrapper around that to replace the current radio buttons called `<RadioList />` which just pre-sets options for the table component (since it assumes it will only have one column of values and will take an array of `{ name, value }` objects as the `data` prop).

mccalluc commented on src/components/genes/Genes.js:
```diff
@@ -1,49 +1,34 @@
+  const data = Object.entries(genesSelected).sort(
+    (a, b) => a[0].localeCompare(b[0]),
+  ).map(
+    ([name, value]) => ({ name, value }),
```
(I see that the sort and localeCompare are carried over, but we don't have those on factors... Unless the motivation for the difference is clear to you, can file an issue asking that it be unified?)

keller-mark replied:
I am not sure, but maybe the difference is that factors should remain in its initial ordering because there are fewer entries ("cluster" and "subcluster") and they may correspond to levels of a hierarchy? Whereas for genes, there may be hundreds, and an alphabetical ordering makes more sense?

mccalluc commented on src/components/selectable-table/SelectableTable.js:
```diff
@@ -0,0 +1,155 @@
+
+  // Callback function to update the `selectedRows` state.
+  const onSelectRow = useCallback((value, checked) => {
+    if (!allowMultiple) {
+      if (checked) {
+        setSelectedRows([value]);
+      } else if (!checked && allowUncheck) {
+        setSelectedRows([]);
+      }
+    } else if (checked) {
+      setSelectedRows(union(selectedRows || [], [value]));
+    } else if (!checked && allowUncheck) {
+      setSelectedRows(difference(selectedRows || [], [value]));
+    }
```
```suggestion
    if (!allowMultiple) {
      if (checked) {
        setSelectedRows([value]);
      } else if (allowUncheck) {
        setSelectedRows([]);
      }
    } else {
      if (checked) {
        setSelectedRows(union(selectedRows || [], [value]));
      } else if (allowUncheck) {
        setSelectedRows(difference(selectedRows || [], [value]));
      }
    }
```

- In the `else` you know `!checked` will be true, so I don't think it's necessary.
- Perhaps wrap the second part in a single `else` to emphasize the parallels?
- I'm wondering about the case where neither `checked` nor `allowUncheck` is true... You could simplify the inner if-then-else, by adding a check outside:

```suggestion
  if (checked || allowUncheck) {
    if (!allowMultiple) {
      setSelectedRows(checked ? [value] : []);
    } else {
      setSelectedRows(
        checked
          ? union(selectedRows || [], [value])
          : difference(selectedRows || [], [value])
      );
    }
  }
``` 

I'm not sure this is an improvement...

keller-mark replied:
I like your second suggested change, and will adopt that!

mccalluc commented on src/components/selectable-table/SelectableTable.js:
```diff
@@ -0,0 +1,155 @@
+    // Check whether an initial set of rows should be selected.
+    const initialSelectedRows = data
+      .map((d) => {
+        if (d[valueKey] !== undefined && d[valueKey]) {
```
`undefined` is false-y, so why not just `d[valueKey]`?

keller-mark replied:
True!

mccalluc commented on src/components/selectable-table/SelectableTable.js:
```diff
@@ -0,0 +1,155 @@
+  useEffect(() => {
+    // Check whether an initial set of rows should be selected.
+    const initialSelectedRows = data
+      .map((d) => {
+        if (d[valueKey] !== undefined && d[valueKey]) {
+          return d[idKey];
+        }
+        return null;
+      })
```
For that matter, why not:
```suggestion
      .map((d) => d[valueKey] && d[idKey])
```

keller-mark replied:
I like the if statement for readability since sometimes using the `&&` to produce some "return value" takes me a second to understand what is going on. But maybe that is just me.

mccalluc commented on src/components/selectable-table/SelectableTable.js:
```diff
@@ -0,0 +1,155 @@
+      return;
+    }
+    const selectedRowData = getDataFromIds(selectedRows);
+    if (!allowMultiple && selectedRows.length === 1) {
+      onChange(selectedRowData[0]);
+    } else if (!allowMultiple && selectedRows.length === 0) {
+      onChange(null);
+    } else if (allowMultiple) {
+      onChange(selectedRowData);
+    }
```
```suggestion
    if (allowMultiple) {
      onChange(selectedRowData);
    } else if (selectedRows.length === 1) {
      onChange(selectedRowData[0]);
    } else if (selectedRows.length === 0) {
      onChange(null);
    }
```

I might be missing something.

keller-mark replied:
No this looks correct, thanks!

mccalluc commented on src/css/_colors.scss:
```diff
@@ -35,6 +36,7 @@ $colors-light: (
 
+  "gray-light": #D3D3D3,
   "gray-mid": #808080,
   "gray-dark": #555555
```
Maybe a mixin for the greys that are the same in both?

keller-mark replied:
Updated in https://github.com/hubmapconsortium/vitessce/pull/510/commits/b9c2dacd0acbdcf5766d3f89ffda9b01e234aa4c

keller-mark commented on src/components/channels/index.js:
```diff
@@ -1 +1 @@
-export ChannelsSubscriber from './ChannelsSubscriber';
+export { default as ChannelsSubscriber } from './ChannelsSubscriber';
```
It threw an error when I tried to use that syntax. Also the change here is consistent with all of the other exports in vitessce, for example https://github.com/hubmapconsortium/vitessce/blob/master/src/components/scatterplot/index.js

keller-mark replied:
No one currently uses this ChannelsSubscriber export which is probably why this has not been caught previously.

mccalluc commented:
I don't see any change in the UI when I launch locally... Is there something else I should do to see it?

keller-mark commented:
If you are seeing <Genes /> or <Factors /> then you should already be seeing it. It is just that the components now look like this

````
<Genes>
    <RadioList>
        <SelectableTable />
    </RadioList>
</Genes>
````

and the `RadioList` component ensures that the SelectableTable uses radio button logic for making selections, and only has one column of data.

The table is styled in such a way that this configuration results in a UI that is identical to the radio buttons that were previously there.

mccalluc commented:
> The table is styled in such a way that this configuration results in a UI that is identical to the radio buttons that were previously there.

Ok, that makes sense, but I think it's a design goal for Nils to actually get rid of the radio buttons? I believe #491 captured his intent. If you don't have the same impression, can you ask him to clarify? (Since it can be hard to get his time, you might also consider asking Tiffany, or Sehi as a fallback: They both have a better sense for UX issues than I do.)

keller-mark commented:
I understand that the goal is to get rid of the buttons, but given that there is currently no alternative to the Genes and Factors components, I added the RadioList component as a simple wrapper that uses consistent styles, but I do intend for the RadioList component to be temporary. 

The actual `SelectableTable` that I implemented does not look like the radio buttons when it is used on its own. Apologies for not making this clear in the original PR description. I have made a new branch with a demo here https://github.com/hubmapconsortium/vitessce/compare/keller-mark/selectable-table...keller-mark/selectable-table-demo

And here are some screenshots:

when `allowMultiple` is false:
<img width="1320" alt="Screen Shot 2020-04-21 at 4 02 37 PM" src="https://user-images.githubusercontent.com/7525285/79908659-a067d780-83e9-11ea-8314-feaf80644e4b.png">

when `allowMultiple` is true:

<img width="1320" alt="Screen Shot 2020-04-21 at 4 01 32 PM" src="https://user-images.githubusercontent.com/7525285/79908688-acec3000-83e9-11ea-8fed-7c3a79757b83.png">

hover:
<img width="1320" alt="Screen Shot 2020-04-21 at 4 01 26 PM" src="https://user-images.githubusercontent.com/7525285/79908700-b1184d80-83e9-11ea-9921-d9e0bc9f8a84.png">

checked:
<img width="1320" alt="Screen Shot 2020-04-21 at 4 01 21 PM" src="https://user-images.githubusercontent.com/7525285/79908711-b6759800-83e9-11ea-94dd-0ff387249b5b.png">

keller-mark commented:
Whoops, in the screenshots above I was not calling the `onReady` function so the other components were not loading properly. Here is what it looks like with those working:

<img width="1320" alt="Screen Shot 2020-04-21 at 4 16 56 PM" src="https://user-images.githubusercontent.com/7525285/79909813-9c3cb980-83eb-11ea-80b2-36d2266539a8.png">

mccalluc commented:
I don't want to block this, but my memory of Nils' preferences was that he really just didn't want to see radio buttons, period. I thin this PR is great, but I just don't think we should close #491 yet, while there are still radio buttons, even if they are inside a table. Does that make sense?

keller-mark commented:
Sure, that makes sense. I will update this to get rid of the `RadioList` wrapper component.

keller-mark commented:
New screenshot
<img width="1531" alt="Screen Shot 2020-04-21 at 5 39 55 PM" src="https://user-images.githubusercontent.com/7525285/79916705-2b9b9a00-83f7-11ea-8276-6f0fa32b87ec.png">

mccalluc commented:
- Changed "Fixes" to "Towards" up at the top so #491 doesn't auto-close.
- Are you waiting for more feedback from me?
- Looking at the last screenshot: Has the padding around the list increased? The frames themselves take up a good amount of space, so I think we'll need to keep pretty tight margins.

keller-mark commented:
No worries, and I will reduce the padding.

By non-radio button do you mean that you want to be able to select multiple rows of Factors and Genes? Or just that the icon should be a square rather than a circle? 

In the table component right now, all selections are implemented using checkbox input elements, but I style the icons as circles to make it more clear that only one row can be selected if the `allowMultiple` prop is true. If `allowMultiple` is false, then the icons have a smaller border radius making them look more like checkboxes.

mccalluc commented:
> By non-radio button do you mean that you want to be able to select multiple rows of Factors and Genes? Or just that the icon should be a square rather than a circle?

Sorry not to be more clear -- What I think Nils wants is _no_ radiobutton _or_ checkbox ... the row itself should be selected on click (probably with a change in background color) something like a spreadsheet.

keller-mark commented:
Oooh that makes much more sense now. Apologies for not grasping that before.

keller-mark commented:
New screenshot. There padding you see now between the table row separator lines and the edge of the card is due to the existing `padding: 1.25rem` on the `.card-body` class. I personally would be in favor of reducing this to `1.0rem` but I also don't think it is a big deal either way.
 
![Screen Shot 2020-04-24 at 11 55 22 AM](https://user-images.githubusercontent.com/7525285/80232248-94b32580-8622-11ea-9471-d7fb6881a039.png)
